### PR TITLE
Departure releases

### PR DESCRIPTION
**Pull request by AndyTWF**
- Allow controllers to request a departure release
- Allow requested controllers to approve a departure release
- Allow requested controllers to reject a departure release
- Allow requested controllers to acknowledge a departure release

### CODE REVIEW

File: app/Http/Controllers/DepartureReleaseController.php [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        );
+
+        return response()->json(['id' => $releaseId], 201);
+    }
+
+    public function approveReleaseRequest(
+        Request $request,
+        int $id
+    ): JsonResponse {
+        $departureReleaseRequest = DepartureReleaseRequest::findOrFail($id);
```
Comment (AxonC) [inline app/Http/Controllers/DepartureReleaseController.php:68]: This I think is an uncaught exception

Comment (AndyTWF) [reply app/Http/Controllers/DepartureReleaseController.php:68]: It is - Laravel auto converts it to a 404 :)

**Review by AxonC — APPROVED** [review]

File: app/Http/Controllers/DepartureReleaseController.php [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                        if (!ControllerPosition::where('id', $value)->canRequestDepartureReleases()->exists()) {
+                            $fail(sprintf('Controller position %d cannot request departure releases', $value));
+                        }
+                    },
+                    'not_in:target_controller_ids.*',
+                ],
+                'target_controller_id' => [
+                    'required',
+                    'integer',
+                    function ($attribute, $value, $fail) {
```
Comment (AxonC) [inline app/Http/Controllers/DepartureReleaseController.php:44]: Can this be its own rule object perhaps?

File: app/Http/Controllers/DepartureReleaseController.php [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            $responseCode = 200;
+        } catch (DepartureReleaseDecisionNotAllowedException $decisionNotAllowedException) {
+            Log::warning(
+                sprintf(
+                    'User %d attempted to cancel release %d without permission',
+                    Auth::id(),
+                    $departureReleaseRequest->id
+                )
+            );
+            $responseCode = 422;
```
Comment (AxonC) [inline app/Http/Controllers/DepartureReleaseController.php:186]: Hmmmmm just thinking - but to be honest this only affects you anyway - the 422 status code will have multiple paths via the validator and further via this error here. Is it possibly best served to use something like a 403?

File: app/Providers/RouteServiceProvider.php [hunk 4]
```diff
@@ -29,6 +29,12 @@ class RouteServiceProvider extends ServiceProvider
      */
     public const HOME = 'horizon';
 
+    public function boot()
+    {
+        parent::boot();
+        Route::model('departureReleaseRequest', DepartureReleaseRequest::class);
```
Comment (AxonC) [inline app/Providers/RouteServiceProvider.php:35]: I'm assuming the model name doesn't match well in the routes file for this?

File: app/Http/Controllers/DepartureReleaseController.php (hunk 2 above)
Comment (AxonC) [reply app/Http/Controllers/DepartureReleaseController.php:44]: Or maybe even middleware for this request?

**Review by AxonC — CHANGES_REQUESTED** [review]

File: app/Providers/RouteServiceProvider.php (hunk 4 above)
Comment (AndyTWF) [reply app/Providers/RouteServiceProvider.php:35]: Yeah I think it didnt like this when I was first setting up, fixed now.

File: app/Http/Controllers/DepartureReleaseController.php (hunk 3 above)
Comment (AndyTWF) [reply app/Http/Controllers/DepartureReleaseController.php:186]: Returning 403 now :)

File: app/Http/Controllers/DepartureReleaseController.php (hunk 2 above)
Comment (AndyTWF) [reply app/Http/Controllers/DepartureReleaseController.php:44]: Done!

**Review by AxonC — APPROVED** [review]

**Pull request merged**